llvm/tools/sycl-post-link/CompileTimePropertiesPass.cpp

```diff
@@ -142,9 +152,124 @@ PreservedAnalyses CompileTimePropertiesPass::run(Module &M,
+
+  // Get the global variable with the annotation string.
+  const GlobalVariable *AnnotStrArgGV = nullptr;
+  if (auto *GEP = dyn_cast<GEPOperator>(IntrInst->getArgOperand(1)))
+    if (auto *C = dyn_cast<Constant>(GEP->getOperand(0)))
+      AnnotStrArgGV = dyn_cast<GlobalVariable>(C);
+  if (!AnnotStrArgGV)
+    return false;
+
+  // We only need to consider annotations with "sycl-properties" annotation
+  // string.
+  Optional<StringRef> AnnotStr = getGlobalVariableString(AnnotStrArgGV);
+  if (!AnnotStr || AnnotStr->str() != "sycl-properties")
+    return false;
+
+  // Read the annotation values and create the new annotation string.
+  std::string NewAnnotString = "";
+  if (const auto *Cast =
+          dyn_cast<BitCastOperator>(IntrInst->getArgOperand(4))) {
+    if (const auto *AnnotValsGV =
+            dyn_cast<GlobalVariable>(Cast->getOperand(0))) {
+      if (const auto *AnnotValsAggr =
+              dyn_cast<ConstantAggregate>(AnnotValsGV->getInitializer())) {
+        assert(
+            (AnnotValsAggr->getNumOperands() & 1) == 0 &&
+            "sycl-properties annotation must have an even number of annotation "
+            "values.");
+
+        // Iterate over the pairs of property meta-names and meta-values.
+        for (size_t I = 0; I < AnnotValsAggr->getNumOperands(); I += 2) {
+          Optional<StringRef> PropMetaName =
+              getGlobalVariableString(AnnotValsAggr->getOperand(I));
+          Optional<StringRef> PropMetaValue =
+              getGlobalVariableString(AnnotValsAggr->getOperand(I + 1));
+
+          assert(PropMetaName &&
+                 "Unexpected format for property name in annotation.");
+
+          auto DecorIt = SpirvDecorMap.find(*PropMetaName);
+          if (DecorIt == SpirvDecorMap.end())
+            continue;
+          uint32_t DecorCode = DecorIt->second.Code;
+
+          // Expected format is '{X}' or '{X:Y}' where X is decoration ID and
+          // Y is the value if present. It encloses Y in " to ensure that
+          // string values are handled correctly. Note that " around values are
+          // always valid, even if the decoration parameters are not strings.
+          NewAnnotString += "{" + std::to_string(DecorCode);
+          if (PropMetaValue)
+            NewAnnotString += ":\"" + PropMetaValue->str() + "\"";
+          NewAnnotString += "}";
+        }
+      }
+    }
+  }
+
+  // If the new annotation string is empty there is no reason to keep it, so
+  // replace it with the first operand and mark it for removal.
+  if (NewAnnotString == "") {
+    IntrInst->replaceAllUsesWith(IntrInst->getOperand(0));
+    RemovableAnnotations.push_back(IntrInst);
+    return true;
+  }
+
+  // Either reuse a previously generated one or create a new global variable
+  // with the new annotation string.
+  GlobalVariable *NewAnnotStringGV = nullptr;
+  auto ExistingNewAnnotStringIt = NewAnnotationStrings.find(NewAnnotString);
```

Comment:
Some difference in naming: `NewAnnotationStrings` but `NewAnnotString` or `ExistingNewAnnotStringIt`.